### ISSUE TITLE
BatchNormalization bug when used in generator of a GAN

### ISSUE THREAD

**Issue by mjdietzx** [fchollet/keras#5644]
Easy to reproduce this bug when training GANs but probably occurs in other use cases as well. When `BatchNormalization` is used in the `generator` of a GAN, `combined.train_on_batch` fails. It's really weird but for some reason `combined.train_on_batch` does not calculate the loss correctly when batch norm is used in the generator. I tested the loss by hand doing something like:

```python
loss = combined.train_on_batch(x, y)
combined_pred = combined.predict(x)
loss_check = K.eval(custom_loss(y, combined_pred))
>>> loss != loss_check
```

You can reproduce this in `wGAN` branch here: https://github.com/wayaai/GAN-Sandbox and uncomment the BatchNorm layer in the generator.

**Comment by unrealwill** [comment]
Hello,

Probably a duplicate of : 
https://github.com/fchollet/keras/issues/5576

**Comment by mjdietzx** [comment]
I don't think these issues are duplicated b/c this issue is when BN is used in generator and it seems to mess up the computation of the loss. The issue you referenced is when BN is used in discriminator and causes an expection to be raised b/c of problem in the computation graph. 

this problem is really hard to discover and may be causing big problems behind the scenes without people realizing. I'm not sure if `predict_on_batch` and `train_on_batch` use different batch normalization statistics which could explain a small difference but this difference in loss seems to really blow up as training progresses which makes me think there is a big problem somewhere.

I added test cases that reproduce this and show exactly what is going on:
https://github.com/fchollet/keras/pull/5647

**Comment by unrealwill** [comment]
I'm not sure your test shows something.
Models can have some regularization terms which will be added to the loss of the model, but obviously not added when you compute a custom_loss on the output of a model.

Also when you use predict the "learningPhase" variable is not the same as when in training, so the behaviour is expected to be different.

https://github.com/fchollet/keras/blob/master/keras/engine/training.py#L1265

Keras is currently in a PR freeze. 
Maybe some correction for Batch Normalization have already been made into the keras2 branch.

**Comment by mjdietzx** [comment]
I use keras-2 and see this problem in both versions. directed PR at keras-2.

I'm not using any regularization (in my code where I see error) or in the tests so that shouldn't be a prob. I quickly looked through the code and it doesn't look like `learningPhase` should change the behavior of the simple models i included in tests (just a dense layer, batch norm, a dense layer and sigmoid activation). maybe something is going on that im missing?

im hoping this test highlights the error im seeing. in my case i have a custom loss and more complicated models and the loss during training is very obviously wrong while if i manually compute the loss as done in tests it is correct (and as training progresses this difference increases a lot)

**Comment by unrealwill** [comment]
I don't know which mode you use for batchNormalization but in mode 0 (default mode)
learningPhase has an impact : 
https://github.com/fchollet/keras/blob/master/keras/layers/normalization.py#L143

**Comment by hadarpo** [comment]
I think that this problem does not depend on the specific model and is not due to regularization. Perhaps the reason is that batch normalization needs an epoch context in order to calculate statistics for normalization and hence does not behaves properly with train_on_batch (rather than fit or fit_generator)

**Comment by oliblum90** [comment]
I have the same problem. Is there a solution yet?

**Comment by HitLuca** [comment]
Can confirm the issue, in order to fix this problem you can clone the BatchNormalizationClass and in the call() method change 
`return K.in_train_phase(normed_training, normalize_inference, training=training)`
into
`return K.in_train_phase(normed_training, normalize_inference, training=True)`
